token/services/ttx/convert.go

```diff
@@ -0,0 +1,271 @@
+	options, err := CompileServiceOptions(opts...)
+	if err != nil {
+		return nil, nil, errors.WithMessagef(err, "failed to compile options")
+	}
+	resultBoxed, err := context.RunView(NewRequestConversionView(
+		issuer,
+		wallet,
+		tokens,
+		notAnonymous,
+	).WithWallet(wallet).WithTMSID(options.TMSID()).WithRecipientData(recipientData))
+	if err != nil {
+		return nil, nil, err
+	}
+	result := resultBoxed.([]interface{})
+	ir := result[0].(*ConversionRequest)
+	return ir.RecipientData.Identity, result[1].(view.Session), nil
+}
+
+func (r *ConversionInitiatorView) Call(context view.Context) (interface{}, error) {
+	span := context.StartSpan("conversion_request_view")
+	defer span.End()
+	logger.Debugf("Respond request recipient identity using wallet [%s]", r.Wallet)
+
+	span.AddEvent("start_session")
+	session, err := session.NewJSON(context, context.Initiator(), r.Issuer)
+	if err != nil {
+		logger.Errorf("failed to get session to [%s]: [%s]", r.Issuer, err)
+		return nil, errors.Wrapf(err, "failed to get session to [%s]", r.Issuer)
+	}
+
+	// first agreement
+	agreement := &ConversionAgreement{}
+	span.AddEvent("send_conversion_agreement")
+	err = session.SendWithContext(context.Context(), agreement)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to send recipient data")
+	}
+
+	if err := session.ReceiveWithTimeout(agreement, 1*time.Minute); err != nil {
+		return nil, errors.Wrapf(err, "failed to receive conversion agreement")
+	}
+
+	// prepare request
+	span.AddEvent("send_conversion_request")
+
+	// - recipient identity
+	tmsID, recipientIdentity, auditInfo, tokenMetadata, err := r.getRecipientIdentity(context)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get recipient identity")
+	}
+	// - proof
+	tms := token.GetManagementService(context, token.WithTMSID(*tmsID))
+	if tms == nil {
+		return nil, errors.Errorf("tms not found for [%s]", tmsID)
+	}
+	proof, err := tms.TokensService().GenConversionProof(agreement.Challenge, r.Tokens)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate proof")
+	}
+
+	wr := &ConversionRequest{
+		ID:    agreement.Challenge,
+		TMSID: *tmsID,
+		RecipientData: RecipientData{
+			Identity:      recipientIdentity,
+			AuditInfo:     auditInfo,
+			TokenMetadata: tokenMetadata,
+		},
+		Tokens:       r.Tokens,
+		Proof:        proof,
+		NotAnonymous: r.NotAnonymous,
+	}
+	err = session.SendWithContext(context.Context(), wr)
+	if err != nil {
+		logger.Errorf("failed to send recipient data: [%s]", err)
+		return nil, errors.Wrapf(err, "failed to send recipient data")
+	}
+
+	return []interface{}{wr, session.Session()}, nil
+}
+
+// WithWallet sets the wallet to use to retrieve a recipient identity if it has not been passed already
+func (r *ConversionInitiatorView) WithWallet(wallet string) *ConversionInitiatorView {
```

Comment:
It is a bit confusing to have two strategies to pass arguments: `With...` and passing them to the constructor. We could either have `opts` or pass them directly. In any case I would force to pass them on instantiation to enforce immutability. Otherwise one might think that WithWallet is meant to be called also after instantiation.

Reply:
indeed, they can also after initialization